_posts/2021-03-05-automatically-generated-api-tests-with-clojure-and-reitit.md

```diff
@@ -0,0 +1,332 @@
+be used to find the created resource.
+
+```
+$ curl -d "Hello" localhost:3001/api/resource   
+{:cause :missing-authentication}
+$ curl -H "Auth: 1" -H "Content-Type: application/json" -d '{"content": "Hello"}' localhost:3001/api/resource
+{:resource-id "4"}
+$ curl -H "Auth: 1" localhost:3001/api/resource/4
+{:content "Hello"}
+```
+
+Using a different Auth id than the one used to create the resource initially
+returns an error, but if the original user is used to add the other user into
+the resource, the other user can also access the resource.
+
+```
+$ curl -H "Auth: 2" localhost:3001/api/resource/4
+{:cause :not-authorized}
+$ curl -H "Auth: 1" -H "Content-Type: application/json" -d '{"resource-id": "4", "token": "2"}' localhost:3001/api/authorization
+{}%                                 
+$ curl -H "Auth: 2" localhost:3001/api/resource/4
+{:content "Hello"}
+```
+
+The API looks like the following piece of code:
+
+```clojure
+(def api
+  [["/api"
+   ["/authorization"
+    {:post   add-authorization
+     :delete remove-authorization}]
+   ["/resource"
+    ["" {:post create-resource}]
+    ["/:resource-id"
+     {:get    read-resource
+      :put    update-resource
+      :delete delete-resource}]]]])
+```
+
+Testing
+=======
+
+Traditionally testing an endpoint like this is relatively simple. Just make some requests and
+check that the returned value is the expected one.
+
+```clojure
+(defn- request
+  ([method token uri]
+   (request method token uri nil))
+  ([method token uri body]
+   (-> (handler/app
+         {:uri            uri
+          :request-method method
+          :headers        {"auth" token}
+          :body-params    body})
+       (select-keys [:status :body])
+       (update :body read-string))))
+
+(deftest test-resource
+  (let [resource-id (-> (request :post 1 "/api/resource" {:content "foo"}) :body :resource-id)]
+    (is (= {:status 200 :body {:content "foo"}}
+           (request :get 1 (str "/api/resource/" resource-id))))
+    (is (= {:status 404 :body {:cause :not-found}}
+           (request :get 1 (str "/api/resource/foo"))))
+    (is (= {:status 400 :body {:cause :missing-authentication}}
+           (request :get nil (str "/api/resource/" resource-id))))
+    (is (= {:status 403 :body {:cause :not-authorized}}
+           (request :get 2 (str "/api/resource/" resource-id))))
+    (is (= {:status 200 :body {}}
+           (request :post 1 "/api/authorization" {:resource-id resource-id :token 2})))
+    (is (= {:status 200 :body {:content "foo"}}
+           (request :get 2 (str "/api/resource/" resource-id))))))
+```
+
+However, what happens if you accidentally skip testing some endpoint? Normally
+you just have to wish that the issue is spotted during code review. If you do
+some larger refactoring, it's possible that both the developer and the reviewer
+miss that some parts of the software aren't tested. Clojure and Reitit makes it
```

Comment:
I think a new paragraph break before "Clojure & reitit makes..." would make it flow better. In it's current form that change of topic comes out of the blue and it is difficult to read.